File: ws/chiriboga/u1/ws08CDJavaClasses/ChickenFarmSimulatorPython/farm_simulator/utils/FileManager.py
# File: utils/FileManager.py

import json
from datetime import datetime
from farm_simulator.model import Chicken, ChickenCoop, ChickenFarmer

FILE_PATH = "ChickenFarmSimulator.json"

def create_chicken(scanner, date_format):
    chicken = get_chicken_details(scanner, date_format)

    print("Do you want to add this chicken to an existing coop? (yes/no): ")
    add_to_coop = scanner.nextLine().strip().lower()

    if add_to_coop == "yes":
        farmers = read_all_farmers()
        print("Available Farmers:")
        for farmer in farmers:
            print(farmer)
        print("Enter Chicken Farmer ID to add the chicken to: ")
        farmer_id = int(scanner.nextLine())
        farmer = find_farmer(farmer_id)

        if farmer:
            print("Available Coops:")
            for coop in farmer.coops:
                print(coop)
            print("Enter Chicken Coop ID to add the chicken to: ")
            coop_id = int(scanner.nextLine())
            for coop in farmer.coops:
                if coop.id == coop_id:
                    coop.add_chicken(chicken)
                    update_farmer(farmer)
                    print("Chicken added to existing coop successfully!")
                    return
            print("Chicken Coop not found.")
        else:
            print("Chicken Farmer not found.")
    else:
        print("Enter Chicken Coop ID: ")
        coop_id = int(scanner.nextLine())
        chickens = [chicken]
        chicken_coop = ChickenCoop(coop_id, chickens)

        print("Enter Chicken Farmer ID: ")
        farmer_id = int(scanner.nextLine())

        print("Enter Chicken Farmer Name: ")
        farmer_name = scanner.nextLine()

        coops = [chicken_coop]
        chicken_farmer = ChickenFarmer(farmer_id, farmer_name, coops)

        create_farmer(chicken_farmer)
        print("Chicken created successfully!")
        print(chicken_farmer)

def read_chickens():
    print("All Chickens: ")
    for farmer in read_all_farmers():
        print(farmer)

def update_chicken(scanner, date_format):
    print("Enter Chicken Farmer ID to update: ")
    farmer_id = int(scanner.nextLine())
    farmer = find_farmer(farmer_id)
    if farmer:
        print("Enter Chicken ID to update: ")
        chicken_id = int(scanner.nextLine())
        updated_chicken = get_chicken_details(scanner, date_format)
        updated_chicken.id = chicken_id
        for coop in farmer.coops:
            for i in range(len(coop.chickens)):
                if coop.chickens[i].id == chicken_id:
                    coop.chickens[i] = updated_chicken
                    update_farmer(farmer)
                    print("Chicken updated successfully!")
                    return
        print("Chicken not found in the specified farmer's coops.")
    else:
        print("Chicken Farmer not found.")

def delete_chicken(scanner):
    print("Enter Chicken Farmer ID to delete chicken from: ")
    farmer_id = int(scanner.nextLine())
    farmer = find_farmer(farmer_id)
    if farmer:
        print("Enter Chicken ID to delete: ")
        chicken_id = int(scanner.nextLine())
        for coop in farmer.coops:
            coop.chickens = [chicken for chicken in coop.chickens if chicken.id != chicken_id]
            update_farmer(farmer)
            print("Chicken deleted successfully!")
            return
        print("Chicken not found in the specified farmer's coops.")
    else:
        print("Chicken Farmer not found.")

def find_chicken(scanner):
    print("Enter Chicken Farmer ID to find chicken in: ")
    farmer_id = int(scanner.nextLine())
    farmer = find_farmer(farmer_id)
    if farmer:
        print("Enter Chicken ID to find: ")
        chicken_id = int(scanner.nextLine())
        for coop in farmer.coops:
            for chicken in coop.chickens:
                if chicken.id == chicken_id:
                    print(f"Found Chicken: {chicken}")
                    return
        print("Chicken not found in the specified farmer's coops.")
    else:
        print("Chicken Farmer not found.")

def get_chicken_details(scanner, date_format):
    print("Enter chicken data:")
    print("ID: ")
    id = int(scanner.nextLine())

    print("Name: ")
    name = scanner.nextLine()

    print("Color: ")
    color = scanner.nextLine()

    print("Molting (true/false): ")
    molting = scanner.nextLine().strip().lower() == "true"

    born_on_date = None
    while born_on_date is None:
        print("Born on Date (yyyy-MM-dd): ")
        born_on_date_str = scanner.nextLine()

        try:
            born_on_date = datetime.strptime(born_on_date_str, "%Y-%m-%d").date()
        except ValueError:
            print("Invalid date format. Please use yyyy-MM-dd")

    return Chicken(id, name, color, molting, born_on_date)

def create_farmer(farmer):
    farmers = read_all_farmers()
    if not farmers:
        farmers = []
    farmers.append(farmer)
    save_all_farmers(farmers)

def read_all_farmers():
    try:
        with open(FILE_PATH, 'r') as file:
            farmers_data = json.load(file)
            farmers = [ChickenFarmer(f['id'], f['name'], [ChickenCoop(c['id'], [Chicken(ch['id'], ch['name'], ch['color'], ch['molting'], datetime.strptime(ch['born_on_date'], "%Y-%m-%d").date()) for ch in c['chickens']]) for c in f['coops']]) for f in farmers_data]
            return farmers
    except IOError:
        return []

def find_farmer(id):
    farmers = read_all_farmers()
    for farmer in farmers:
        if farmer.id == id:
            return farmer
    return None

def update_farmer(updated_farmer):
    farmers = read_all_farmers()
    for i in range(len(farmers)):
        if farmers[i].id == updated_farmer.id:
            farmers[i] = updated_farmer
            break
    save_all_farmers(farmers)

def delete_farmer(id):
    farmers = read_all_farmers()
    farmers = [farmer for farmer in farmers if farmer.id != id]
    save_all_farmers(farmers)

def save_all_farmers(farmers):
    with open(FILE_PATH, 'w') as file:
        farmers_data = [{'id': farmer.id, 'name': farmer.name, 'coops': [{'id': coop.id, 'chickens': [{'id': chicken.id, 'name': chicken.name, 'color': chicken.color, 'molting': chicken.molting, 'born_on_date': chicken.born_on_date.strftime ("%Y-%m-%d")} for chicken in coop.chickens]} for coop in farmer.coops]} for farmer in farmers] 
        json.dump(farmers_data, file, indent=4)

File: ws/chiriboga/u1/ws08CDJavaClasses/ChickenFarmSimulator/src/utils/FileManager.java
package utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import ec.edu.espe.farmsimulator.model.Chicken;
import ec.edu.espe.farmsimulator.model.ChickenCoop;
import ec.edu.espe.farmsimulator.model.ChickenFarmer;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Scanner;

public class FileManager {

    private static final String FILE_PATH = "ChickenFarmSimulator.json";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    public static void createChicken(Scanner scanner, SimpleDateFormat dateFormat) throws IOException {
        Chicken chicken = getChickenDetails(scanner, dateFormat);

        System.out.println("Do you want to add this chicken to an existing coop? (yes/no): ");
        String addToCoop = scanner.nextLine().trim().toLowerCase();

        if (addToCoop.equals("yes")) {
            List<ChickenFarmer> farmers = FileManager.readAllFarmers();
            System.out.println("Available Farmers:");
            for (ChickenFarmer farmer : farmers) {
                System.out.println(farmer);
            }
            System.out.println("Enter Chicken Farmer ID to add the chicken to: ");
            int farmerId = scanner.nextInt();
            scanner.nextLine();
            ChickenFarmer farmer = FileManager.findFarmer(farmerId);

            if (farmer != null) {
                System.out.println("Available Coops:");
                for (ChickenCoop coop : farmer.getCoops()) {
                    System.out.println(coop);
                }
                System.out.println("Enter Chicken Coop ID to add the chicken to: ");
                int coopId = scanner.nextInt();
                scanner.nextLine();
                for (ChickenCoop coop : farmer.getCoops()) {
                    if (coop.getId() == coopId) {
                        coop.getChickens().add(chicken);
                        FileManager.updateFarmer(farmer);
                        System.out.println("Chicken added to existing coop successfully!");
                        return;
                    }
                }
                System.out.println("Chicken Coop not found.");
            } else {
                System.out.println("Chicken Farmer not found.");
            }
        } else {
            System.out.println("Enter Chicken Coop ID: ");
            int coopId = scanner.nextInt();
            scanner.nextLine();
            List<Chicken> chickens = new ArrayList<>();
            chickens.add(chicken);
            ChickenCoop chickenCoop = new ChickenCoop(coopId, chickens);

            System.out.println("Enter Chicken Farmer ID: ");
            int farmerId = scanner.nextInt();
            scanner.nextLine();

            System.out.println("Enter Chicken Farmer Name: ");
            String farmerName = scanner.nextLine();

            List<ChickenCoop> coops = new ArrayList<>();
            coops.add(chickenCoop);

            ChickenFarmer chickenFarmer = new ChickenFarmer(farmerId, farmerName, coops);

            FileManager.createFarmer(chickenFarmer);
            System.out.println("Chicken created successfully!");
            System.out.println(chickenFarmer);
        }
    }

    public static void readChickens() throws IOException {
        System.out.println("All Chickens: ");
        for (ChickenFarmer farmer : FileManager.readAllFarmers()) {
            System.out.println(farmer);
        }
    }

    public static void updateChicken(Scanner scanner, SimpleDateFormat dateFormat) throws IOException {
        System.out.println("Enter Chicken Farmer ID to update: ");
        int farmerId = scanner.nextInt();
        scanner.nextLine();
        ChickenFarmer farmer = FileManager.findFarmer(farmerId);
        if (farmer != null) {
            System.out.println("Enter Chicken ID to update: ");
            int chickenId = scanner.nextInt();
            scanner.nextLine();
            Chicken updatedChicken = getChickenDetails(scanner, dateFormat);
            updatedChicken.setId(chickenId);
            for (ChickenCoop coop : farmer.getCoops()) {
                for (int i = 0; i < coop.getChickens().size(); i++) {
                    if (coop.getChickens().get(i).getId() == chickenId) {
                        coop.getChickens().set(i, updatedChicken);
                        FileManager.updateFarmer(farmer);
                        System.out.println("Chicken updated successfully!");
                        return;
                    }
                }
            }
            System.out.println("Chicken not found in the specified farmer's coops.");
        } else {
            System.out.println("Chicken Farmer not found.");
        }
    }

    public static void deleteChicken(Scanner scanner) throws IOException {
        System.out.println("Enter Chicken Farmer ID to delete chicken from: ");
        int farmerId = scanner.nextInt();
        scanner.nextLine();
        ChickenFarmer farmer = FileManager.findFarmer(farmerId);
        if (farmer != null) {
            System.out.println("Enter Chicken ID to delete: ");
            int chickenId = scanner.nextInt();
            scanner.nextLine();
            for (ChickenCoop coop : farmer.getCoops()) {
                Iterator<Chicken> iterator = coop.getChickens().iterator();
                while (iterator.hasNext()) {
                    if (iterator.next().getId() == chickenId) {
                        iterator.remove();
                        FileManager.updateFarmer(farmer);
                        System.out.println("Chicken deleted successfully!");
                        return;
                    }
                }
            }
            System.out.println("Chicken not found in the specified farmer's coops.");
        } else {
            System.out.println("Chicken Farmer not found.");
        }
    }

    public static void findChicken(Scanner scanner) throws IOException {
        System.out.println("Enter Chicken Farmer ID to find chicken in: ");
        int farmerId = scanner.nextInt();
        scanner.nextLine();
        ChickenFarmer farmer = FileManager.findFarmer(farmerId);
        if (farmer != null) {
            System.out.println("Enter Chicken ID to find: ");
            int chickenId = scanner.nextInt();
            scanner.nextLine();
            for (ChickenCoop coop : farmer.getCoops()) {
                for (Chicken chicken : coop.getChickens()) {
                    if (chicken.getId() == chickenId) {
                        System.out.println("Found Chicken: " + chicken);
                        return;
                    }
                }
            }
            System.out.println("Chicken not found in the specified farmer's coops.");
        } else {
            System.out.println("Chicken Farmer not found.");
        }
    }

    private static Chicken getChickenDetails(Scanner scanner, SimpleDateFormat dateFormat) {
        System.out.println("Enter chicken data:");
        System.out.println("ID: ");
        int id = scanner.nextInt();
        scanner.nextLine();

        System.out.println("Name: ");
        String name = scanner.nextLine();

        System.out.println("Color: ");
        String color = scanner.nextLine();

        System.out.println("Molting (true/false): ");
        boolean molting = scanner.nextBoolean();
        scanner.nextLine();

        Date bornOnDate = null;
        while (bornOnDate == null) {
            System.out.println("Born on Date (yyyy-MM-dd): ");
            String bornOnDateStr = scanner.nextLine();

            try {
                bornOnDate = dateFormat.parse(bornOnDateStr);
            } catch (ParseException e) {
                System.err.println("Invalid date format. Please use yyyy-MM-dd");
            }
        }

        return new Chicken(id, name, color, molting, bornOnDate);
    }

    public static void createFarmer(ChickenFarmer farmer) throws IOException {
        List<ChickenFarmer> farmers = readAllFarmers();
        if (farmers == null) {
            farmers = new ArrayList<>();
        }
        farmers.add(farmer);
        saveAllFarmers(farmers);
    }

    public static List<ChickenFarmer> readAllFarmers() throws IOException {
        try (FileReader reader = new FileReader(FILE_PATH)) {
            Type listType = new TypeToken<ArrayList<ChickenFarmer>>() {}.getType();
            return gson.fromJson(reader, listType);
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    public static ChickenFarmer findFarmer(int id) throws IOException {
        List<ChickenFarmer> farmers = readAllFarmers();
        for (ChickenFarmer farmer : farmers) {
            if (farmer.getId() == id) {
                return farmer;
            }
        }
        return null;
    }

    public static void updateFarmer(ChickenFarmer updatedFarmer) throws IOException {
        List<ChickenFarmer> farmers = readAllFarmers();
        for (int i = 0; i < farmers.size(); i++) {
            if (farmers.get(i).getId() == updatedFarmer.getId()) {
                farmers.set(i, updatedFarmer);
                break;
            }
        }
        saveAllFarmers(farmers);
    }

    public static void deleteFarmer(int id) throws IOException {
        List<ChickenFarmer> farmers = readAllFarmers();
        farmers.removeIf(farmer -> farmer.getId() == id);
        saveAllFarmers(farmers);
    }

    private static void saveAllFarmers(List<ChickenFarmer> farmers) throws IOException {
        try (FileWriter writer = new FileWriter(FILE_PATH)) {
            gson.toJson(farmers, writer);
        }
    }
}
